src/inmanta/server/services/compilerservice.py

```diff
@@ -601,13 +602,104 @@ async def request_recompile(
+        of that database transaction. It is required to call `CompileService.notify_compile_request_committed()`
+        right after the transaction commits.
+        """
+        return await self._request_recompile(
+            env,
+            force_update,
+            do_export,
+            remote_id,
+            metadata,
+            env_vars,
+            partial,
+            removed_resource_sets,
+            exporter_plugin,
+            notify_failed_compile,
+            failed_compile_message,
+            connection=connection,
+            in_db_transaction=True,
+        )
+
+    async def notify_compile_request_committed(self, compile_id: uuid.UUID) -> None:
+        """
+        This method must be called when the `request_recompile_in_transaction()` method was called and the associated
+        transaction committed successfully. It must be called right after the transaction committed. If the transaction was
+        aborted, this method most not be called.
+
+        This method is safe with respect to server restarts. If the server restarts after the transaction was committed,
+        but before this method was invoked, the server will automatically recover from this and run the requested compile
+        without any need to call this method.
+        """
+        async with self._queue_count_cache_lock:
+            self._queue_count_cache += 1
+        compile_obj: data.Compile = await data.Compile.get_by_id(compile_id)
+        if compile_obj.started:
+            raise Exception(f"Compile with id {compile_obj.id} was already started at {compile_obj.started}.")
+        await self._queue(compile_obj)
 
+    async def _request_recompile(
+        self,
+        env: data.Environment,
+        force_update: bool,
+        do_export: bool,
+        remote_id: uuid.UUID,
+        metadata: Optional[JsonType] = None,
+        env_vars: Optional[Mapping[str, str]] = None,
+        partial: bool = False,
+        removed_resource_sets: Optional[List[str]] = None,
+        exporter_plugin: Optional[str] = None,
+        notify_failed_compile: Optional[bool] = None,
+        failed_compile_message: Optional[str] = None,
+        in_db_transaction: bool = False,
+        connection: Optional[Connection] = None,
+    ) -> Tuple[Optional[uuid.UUID], Warnings]:
+        """
+        :param in_db_transaction: True iff the given connection is part of an ongoing database transaction.
+        :param connection: An optional database connection. A connection must be provided when in_db_transaction is True.
         """
+        if in_db_transaction and not connection:
+            raise Exception("A connection should be provided when in_db_transaction is True.")
+        if in_db_transaction and connection and not connection.is_in_transaction():
+            raise Exception("in_db_transaction is True, but the given connection is not executing a transaction.")
```

Comment:
Doesn't this make the `in_db_transaction` argument completely redundant? `in_db_transaction = connection is not None and connection.is_in_transaction()`?

Reply:
I would not do that, because the additional argument comes with an additional obligation (call notify).

Reply:
Which would be a clear part of the contract, but you're right that it would be less explicit.